### PR TITLE
Workspace switcher

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Workspace switcher. Only for Focalboard running as a plugin. Intentionally disabled on a personal server.

#### Ticket Link
https://community-daily.mattermost.com/plugins/focalboard/workspace/qgsck6cts3fwpqwyjiupjm5cde/47aa9bb4-6967-4a96-83c7-11bd6b20f1eb/392c8889-6c8f-4cb7-8407-f5c57c1acfe8?c=36f12ec9-0705-4a85-84d1-ea1e7040f2f1

#### Functionality
1. No change in functionality for personal server mode.
2. New workspace switcher is now visible when running Focalboard as a Mattermost plugin.
3. The workspace switcher shows the Focalboard workspaces (named by their corresponding Mattermost channels" and the number of boards in each workspace, for the workspaces the user has access to.
4. If a workspace has no boards, it's not shown in the list.
5. Selecting a workspace from the dropdown takes the user to that workspace.

**Testing note -** please do test on both PostgreSQL and MySQL, and if possible, on MySQL 5.8 as well (current min supported MySQL version).

### CODE REVIEW

File: server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql [hunk 1]
```diff
@@ -12,4 +12,4 @@ CREATE TABLE IF NOT EXISTS {{.prefix}}blocks (
 	update_at BIGINT,
 	delete_at BIGINT,
 	PRIMARY KEY (id, insert_at)
-){{if .mysql}}CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci{{end}};
```
Comment (chenilim) [inline server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: Did we come to a conclusion on what (if any) encoding to set? Per @jespino's last comment in [this thread](https://community.mattermost.com/core/pl/tyr59rq8wb8sifxqgka8gybeqr): "Yes, the problem is, if you not "hardcoded" it, it can use utf8 and everything looks broken, we decided to choose one collation there, the user can still alter the table later. I'm not saying that decision is correct, but not setting it can lead to another kind of problems"

File: server/services/store/sqlstore/workspaces.go [hunk 2]
```diff
@@ -144,3 +145,62 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+	if err != nil {
+		s.logger.Error("ERROR GetUserWorkspaces", mlog.Err(err))
+		return nil, err
+	}
+
+	defer s.CloseRows(rows)
+	return s.userWorkspacesFromRoes(rows)
+}
+
+func (s *SQLStore) userWorkspacesFromRoes(rows *sql.Rows) ([]model.UserWorkspace, error) {
```
Comment (sbishel) [inline server/services/store/sqlstore/workspaces.go:185]: nit: should be  userWorkspacesFromRows

File: server/services/store/sqlstore/workspaces.go [hunk 3]
```diff
@@ -144,3 +145,62 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+		var userWorkspace model.UserWorkspace
+
+		err := rows.Scan(
+			&userWorkspace.ID,
+			&userWorkspace.Title,
+			&userWorkspace.BoardCount,
+		)
+
+		if err != nil {
+			s.logger.Error("ERROR userWorkspacesFromRoes", mlog.Err(err))
```
Comment (sbishel) [inline server/services/store/sqlstore/workspaces.go:198]: same nit.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import {useIntl} from 'react-intl'
+
+import {getUserWorkspaceList} from '../../store/users'
+import {UserWorkspace} from '../../user'
+import {useAppSelector} from '../../store/hooks'
+
+import './workspaceOptions.scss'
+import Search from '../../widgets/icons/search'
+
+const style = {
```
Comment (sbishel) [inline webapp/src/components/workspaceSwitcher/workspaceOptions.tsx:16]: Why style here vs css?

File: webapp/src/pages/boardPage.tsx [hunk 5]
```diff
@@ -206,6 +206,8 @@ const BoardPage = (props: Props) => {
         }
     })
 
+    // }
```
Comment (sbishel) [inline webapp/src/pages/boardPage.tsx:209]: nit - remove.

**Review by sbishel — APPROVED** [review]
Just minor nits.

File: server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql (hunk 1 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: @chenilim both charset and collation can get set in DB connection strings in Focalboard config files. So one doesn't need to hardcode them in here. 

One thing we need to do is document it in Focalboard docs to use `utf8mb4_general_ci` collation or any other `utf8mb4` variant. Can you point me to docs where I can document this?

File: server/services/store/sqlstore/workspaces.go (hunk 2 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go:185]: Done.

File: server/services/store/sqlstore/workspaces.go (hunk 3 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go:198]: Done.

File: webapp/src/pages/boardPage.tsx (hunk 5 above)
Comment (harshilsharma63) [reply webapp/src/pages/boardPage.tsx:209]: Done.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.tsx (hunk 4 above)
Comment (harshilsharma63) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.tsx:16]: Missed this. They are no longer needed. Removed now.

File: server/services/store/sqlstore/workspaces.go [hunk 6]
```diff
@@ -144,3 +145,62 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+
+	if s.dbType == mysqlDBType {
+		query = s.getQueryBuilder().
+			Select("Channels.ID", "Channels.DisplayName", "COUNT(focalboard_blocks.id)").
+			From("focalboard_blocks").
+			Join("ChannelMembers ON focalboard_blocks.workspace_id = ChannelMembers.ChannelId").
+			Join("Channels ON ChannelMembers.ChannelId = Channels.Id").
+			Where("ChannelMembers.UserId = ?", userID).
+			Where("focalboard_blocks.type = 'board'").
+			Where("focalboard_blocks.fields LIKE '%\"isTemplate\":false%'").
```
Comment (jespino) [inline server/services/store/sqlstore/workspaces.go:160]: This is dangerous, somebody can introduce in the board description the text `"isTemplate":false` and this is going to swallow that. Personally I prefer to filter it after the query is returned. At least for MySQL.

File: server/services/store/sqlstore/workspaces.go [hunk 7]
```diff
@@ -144,3 +145,62 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+func (s *SQLStore) GetUserWorkspaces(userID string) ([]model.UserWorkspace, error) {
+	var query sq.SelectBuilder
+
+	if s.dbType == mysqlDBType {
+		query = s.getQueryBuilder().
+			Select("Channels.ID", "Channels.DisplayName", "COUNT(focalboard_blocks.id)").
+			From("focalboard_blocks").
+			Join("ChannelMembers ON focalboard_blocks.workspace_id = ChannelMembers.ChannelId").
+			Join("Channels ON ChannelMembers.ChannelId = Channels.Id").
+			Where("ChannelMembers.UserId = ?", userID).
```
Comment (jespino) [inline server/services/store/sqlstore/workspaces.go:158]: I prefer to use here the `sq.Eq{"ChannelMembers.UserId": userID}` approach, is more explicit.

Comment (jespino) [reply server/services/store/sqlstore/workspaces.go:158]: same in the next line, and in the postgres queries.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        return {
+            label: workspace.title,
+            value: workspace.id,
+            boardCount: workspace.boardCount,
+        }
+    })
+
+    return (
+        <Select
+            classNamePrefix={'WorkspaceOptions'}
```
Comment (jespino) [inline webapp/src/components/workspaceSwitcher/workspaceOptions.tsx:35]: I would expect also the `className` field with `WorkspaceOptions` here.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.scss [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    .boardCount {
+        color: rgba(var(--center-channel-color-rgb), 0.9);
+        font-size: 10px;
+        text-transform: uppercase;
+    }
+}
+
+.WorkspaceOptions__indicators {
```
Comment (jespino) [inline webapp/src/components/workspaceSwitcher/workspaceOptions.scss:31]: I would expect all this to be something like:
```css
.WorkspaceOptions {
    &__indicators{
        ...
    }
    &__control {
        ...
    }
    ...
}
```

File: webapp/src/components/workspaceSwitcher/workspaceOptions.scss [hunk 10]
```diff
@@ -0,0 +1,67 @@
+.workspaceOption {
```
Comment (jespino) [inline webapp/src/components/workspaceSwitcher/workspaceOptions.scss:1]: If workspaceOption is a component, shouldn't it be `.WorkspaceOption`? and, because `.WorkspaceOption` is alwais inside a `WorkspaceOptions` shouldn't it be inside `WorkspaceOptions` class?

File: webapp/src/components/workspaceSwitcher/workspaceOptions.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+const Option = (props: any): JSX.Element => {
+    const {innerProps, innerRef} = props
+
+    return (
+        <div
+            ref={innerRef}
+            {...innerProps}
+            className={`workspaceOption ${props.isFocused ? 'focused' : ''}`}
```
Comment (jespino) [inline webapp/src/components/workspaceSwitcher/workspaceOptions.tsx:82]: Because this is a component, I would expect this to be uppercased class name, like `WorkspaceOption`.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.scss (hunk 10 above)
Comment (jespino) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.scss:1]: To avoid collitions on css when this is running as a plugin we add always the `.focalboard-body` class as prefix, otherwise there are some risk of css classes collition.

Comment (jespino) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.scss:1]: this `.focalboard-body` applies to most of the classes defined in the focalboard css, so it also appies to the other CSS files submitted here.

File: webapp/src/components/workspaceSwitcher/workspaceSwitcher.tsx [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            </div>
+            {
+                showMenu &&
+                <WorkspaceOptions
+                    activeWorkspaceId={props.activeWorkspace.id}
+                    onClose={() => {
+                        setSelectClosed(true)
+                    }}
+                    onChange={(workspaceId: string) => {
+                        const newPath = generatePath(match.path, {...match.params, workspaceId})
```
Comment (jespino) [inline webapp/src/components/workspaceSwitcher/workspaceSwitcher.tsx:48]: This is not totally true. This was this way in other cases because we wanted to preserve the current path, because it can change a lot and there is not an easy way to define what is the right path, in this case, because in shared mode shouldn't be shown, and in personal server shouldn't be shown, you can assume that this url is going to be always `/workspace/:workspaceId`. In any case, if you want to keep the `generatePath` approach, you should remove the boardId and viewId from the current `match.params`, if not you are generating a route that is not correct (boardId and viewId from another workspace).

File: webapp/src/octoClient.ts [hunk 13]
```diff
@@ -381,6 +381,20 @@ class OctoClient {
         }
         return (await this.getJson(response, [])) as IUser[]
     }
+
+    async getUserWorkspaces(): Promise<UserWorkspace[]> {
+        if (!Utils.isFocalboardPlugin()) {
```
Comment (jespino) [inline webapp/src/octoClient.ts:386]: I'm not 100% sure if this is correct, the client library in theory should be usable for third party applications and looks like passing here the `isFocalboardPlugin` is an implementation detail that shouldn't belong to the library.

File: webapp/src/store/users.ts [hunk 14]
```diff
@@ -19,11 +19,12 @@ type UsersStatus = {
     me: IUser|null
     workspaceUsers: {[key: string]: IUser}
     loggedIn: boolean|null
+    userWorkspaces: UserWorkspace[]
```
Comment (jespino) [inline webapp/src/store/users.ts:22]: I would say that this belong to `workspace.ts` not to `users.ts`. `users.ts` handles the users models, in this case, return myself, and the list of users. Whenever you need workspace instances, I expect them to be in `workspace.ts`.

File: webapp/src/widgets/icons/chevronDown.tsx [hunk 15]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react'
+
+import CompassIcon from './compassIcon'
+
+export default function ChevronDown(): JSX.Element {
```
Comment (jespino) [inline webapp/src/widgets/icons/chevronDown.tsx:8]: nit: I personally dislike this kind of icon names, I would prefer something more meaningful. Like the `Search` icon that you also included in this PR.

**Review by jespino — CHANGES_REQUESTED** [review]
In general the approach looks good, I added some comments of things that I think should be change or at least discussed.

File: server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql (hunk 1 above)
Comment (jespino) [reply server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: I'm not 100% comfortable with this change, but I would need to think more about it.

Comment (harshilsharma63) [reply server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: @jespino whats your concern?

File: webapp/src/store/users.ts (hunk 14 above)
Comment (harshilsharma63) [reply webapp/src/store/users.ts]: Done.

File: webapp/src/octoClient.ts (hunk 13 above)
Comment (harshilsharma63) [reply webapp/src/octoClient.ts]: Done.

File: webapp/src/components/workspaceSwitcher/workspaceSwitcher.tsx (hunk 12 above)
Comment (harshilsharma63) [reply webapp/src/components/workspaceSwitcher/workspaceSwitcher.tsx]: Done.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.tsx (hunk 11 above)
Comment (harshilsharma63) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.tsx]: Done.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.scss (hunk 9 above)
Comment (harshilsharma63) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.scss]: Oh, nice! Didn't knew we could do this.

File: webapp/src/components/workspaceSwitcher/workspaceOptions.tsx (hunk 8 above)
Comment (harshilsharma63) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.tsx]: Done.

File: server/services/store/sqlstore/workspaces.go (hunk 7 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go]: Done.

File: server/services/store/sqlstore/workspaces.go (hunk 6 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go]: That's not what happens. If someone types `"isTemplate":false` in UI, it gets saved as `\"isTemplate\":false` (escaped double quotes)

Regular board-
Screenshot 2021-08-23 at 6 33 39 PM

Board with description set as `"isTemplate":false`-
Screenshot 2021-08-23 at 6 34 28 PM

File: webapp/src/components/workspaceSwitcher/workspaceOptions.scss (hunk 10 above)
Comment (harshilsharma63) [reply webapp/src/components/workspaceSwitcher/workspaceOptions.scss]: Have renamed this to just `Options`.

File: webapp/src/widgets/icons/chevronDown.tsx (hunk 15 above)
Comment (harshilsharma63) [reply webapp/src/widgets/icons/chevronDown.tsx:8]: It can be used for multiple purposes such as for dropdown indicators, scroll indicators, etc. The usage doesn't change the fact that it's just a generic, chevron down icon. I prefer to keep this `ChevronDown`.

File: server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql (hunk 1 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: Resolving as we're fine with this approach and adding a default collation.

**Comment by harshilsharma63** [comment]
@asaadmahmood can you please make a round of UI/UX review on this?

**Review by asaadmahmood — CHANGES_REQUESTED** [review]
Have pointed out a bunch of issues to Harshil.
1. A bunch of UI changes are needed.
2. My local instance doesn't load the workspaces.

**Comment by harshilsharma63** [comment]
@asaadmahmood points 1 and 3 are fixed now. I'm working with Jesus and Scott to see if it was an issue with your system or a general issue. WIll update here once tested.

**Review by asaadmahmood — APPROVED** [review]
UI looks good to me!

File: server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.up.sql [hunk 16]
```diff
@@ -0,0 +1,45 @@
+{{if .mysql}}
+SET @targetCollation = (SELECT table_collation from information_schema.tables WHERE table_name = 'Channels') AND table_schema = (SELECT DATABASE());
+
+-- blocks
+SET @updateCollationQuery = CONCAT('ALTER TABLE focalboard_blocks COLLATE ', @targetCollation);
```
Comment (harshilsharma63) [inline server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.up.sql:5]: We need all this to use a variable in the query.

File: server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.up.sql [hunk 17]
```diff
@@ -0,0 +1,58 @@
+{{if .mysql}}
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.up.sql]: Done.

File: webapp/src/components/dialog.tsx [hunk 18]
```diff
@@ -28,7 +30,7 @@ const Dialog = React.memo((props: Props) => {
     useHotkeys('esc', () => props.onClose())
 
     return (
-        <div className='Dialog dialog-back'>
+        <div className={'Dialog dialog-back ' + (props.className || '')}>
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply webapp/src/components/dialog.tsx]: Done.

File: config.json [hunk 19]
```diff
@@ -1,8 +1,8 @@
 {
 	"serverRoot": "http://localhost:8000",
 	"port": 8000,
-	"dbtype": "sqlite3",
-	"dbconfig": "./focalboard.db",
+    "dbtype": "mysql",
```
Comment (jespino) [inline config.json:4]: Why are you changing the defaults here to mysql?

Comment (harshilsharma63) [reply config.json:4]: This was left from my testing. Pushing changes with that removed.

File: server/services/store/sqlstore/migrate.go [hunk 20]
```diff
@@ -145,9 +158,15 @@ func (s *SQLStore) Migrate() error {
 	if err != nil {
 		return err
 	}
+
+	fmt.Println("############################################")
```
Comment (jespino) [inline server/services/store/sqlstore/migrate.go:162]: This looks like debug block. I think you can remove this.

File: server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql [hunk 21]
```diff
@@ -12,4 +12,4 @@ CREATE TABLE IF NOT EXISTS {{.prefix}}blocks (
...
+) {{if .mysql}}DEFAULT CHARACTER SET utf8mb4{{end}};
```
Comment (jespino) [inline server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: Nice. I'm not 100% sure if this solves the problem, but looks like an acceptable solution.

File: server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.down.sql [hunk 22]
```diff
@@ -0,0 +1,2 @@
+-- Nothing to be done here
+-- This page is intentionally left blank
```
Comment (jespino) [inline server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.down.sql:2]: You added an empty query in the `.up.sql` file because if not, it fails, I guess it should be the same here.

File: server/services/store/sqlstore/workspaces.go [hunk 23]
```diff
@@ -145,3 +146,62 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+			GroupBy("Channels.Id", "Channels.DisplayName")
+	} else {
+		query = s.getQueryBuilder().
+			Select("channels.id", "channels.displayname", "count(focalboard_blocks.id)").
+			From("focalboard_blocks").
+			Join("channelmembers ON focalboard_blocks.workspace_id = channelmembers.channelid").
+			Join("channels ON channelmembers.channelid = channels.id").
+			Where(sq.Eq{"channelmembers.userid": userID}).
+			Where(sq.Eq{"focalboard_blocks.type": "board"}).
+			Where("focalboard_blocks.fields ->> 'isTemplate' = 'false'").
```
Comment (jespino) [inline server/services/store/sqlstore/workspaces.go:171]: This is not going to work on SQLite. I guess you have to use the previous block for SQLite.

File: webapp/src/store/workspace.ts [hunk 24]
```diff
@@ -15,9 +17,14 @@ export const fetchWorkspace = createAsyncThunk(
     async () => client.getWorkspace(),
 )
 
+type WorkspaceState = {
+    value: IWorkspace|null
```
Comment (jespino) [inline webapp/src/store/workspace.ts:21]: maybe this should be now `current` instead of `value`.

File: webapp/src/store/workspace.ts [hunk 25]
```diff
@@ -36,6 +44,13 @@ const workspaceSlice = createSlice({
...
 export const {reducer} = workspaceSlice
 
+export const getUserWorkspaces = (state: RootState): UserWorkspace[] => state.workspace.userWorkspaces
+
+export const getUserWorkspaceList = createSelector(
+    getUserWorkspaces,
+    (userWorkspaces) => userWorkspaces,
+)
+
 export function getWorkspace(state: RootState): IWorkspace|null {
```
Comment (jespino) [inline webapp/src/store/workspace.ts:54]: Also, I think this should be `getCurrentWorkspace`.

**Review by jespino — CHANGES_REQUESTED** [review]
Some small changes needed, but otherwise looks good to me.

File: server/services/store/sqlstore/workspaces.go [hunk 26]
```diff
@@ -145,3 +146,62 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+			GroupBy("channels.id", "channels.displayname")
```
[the remark this answers is not in the document]
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go]: Done.

File: config.json (hunk 19 above)
Comment (harshilsharma63) [reply config.json]: Done.

File: server/services/store/sqlstore/migrate.go (hunk 20 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/migrate.go]: Done.

File: server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql (hunk 21 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/migrations/migrations_files/000001_init.up.sql:15]: It did work and I tested on MySQL 8.7 and the 15.

File: server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.down.sql (hunk 22 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/migrations/migrations_files/000011_match_collation.down.sql:2]: Ah, great catch! Fixed now.

File: server/services/store/sqlstore/workspaces.go (hunk 23 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go]: This is only intended to run as a plugin, which doesn't use SQLite.

File: webapp/src/store/workspace.ts (hunk 24 above)
Comment (harshilsharma63) [reply webapp/src/store/workspace.ts]: Done.

File: webapp/src/store/workspace.ts (hunk 25 above)
Comment (harshilsharma63) [reply webapp/src/store/workspace.ts]: Done.

File: server/services/store/sqlstore/workspaces.go [hunk 27]
```diff
@@ -150,30 +156,26 @@ func (s *SQLStore) GetWorkspaceCount() (int64, error) {
...
+	case mysqlDBType:
+		query = query.Where(sq.Like{"focalboard_blocks.fields": "%\"isTemplate\":false%"})
+	case postgresDBType:
+		query = query.Where("focalboard_blocks.fields ->> 'isTemplate' = 'false'")
+	default:
+		return nil, fmt.Errorf("GetUserWorkspaces - %w", errUnsupportedDatabaseError)
 	}
 
-	rows, err := query.Query()
+	query = query.GroupBy("Channels.Id", "Channels.DisplayName")
```
Comment (jespino) [inline server/services/store/sqlstore/workspaces.go:176]: nit: the `GroupBy` can be added to the query in the initial query builder. I don't think is needed to do it later.

**Review by jespino — APPROVED** [review]
Now LGTM. An small nit, but I don't think is strictly necessary.

File: server/services/store/sqlstore/workspaces.go (hunk 27 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/workspaces.go]: Interesting! I didn't know the clauses were order independent! Nice!

**Comment by harshilsharma63** [comment]
@ogi-m this PR is not dev-approved. Can you please test this out?

**Comment by ogi-m** [comment]
Thanks @harshilsharma63, LGTM 🎉 
- workspace switcher visible when running Focalboard as a Mattermost plugin.
- workspace switcher shows the Focalboard workspaces named by their corresponding Mattermost channels and the number of boards in each workspace 
- selecting a workspace from the dropdown takes the user to that workspace.
- only the workspaces the user has access to are shown
- adding/removing a user from a channel adds/removes the channel's workspace to/from the user's workspace switcher
- If a workspace has no boards, it's not shown in the list
- workspaces are added to the switcher when boards are added to the the workspace, and removed if all boards get deleted
- tested on Postgres and MySQL

Note: didn't test on MySQL 5.8, I don't think we have a test server for it but I'll check.

File: webapp/src/components/sidebar/sidebar.scss [hunk 28]
```diff
@@ -127,4 +127,10 @@
         stroke: rgba(var(--sidebar-text-rgb), 0.5);
         stroke-width: 6px;
     }
+
+    @media screen and (min-width: 768px) {
+        .sidebarSwitcher {
+            display: none;
+        }
+    }
```
Comment (kamre) [inline webapp/src/components/sidebar/sidebar.scss:135]: @harshilsharma63 why is this needed? This is a kind of regression for personal server because now it is not possible to hide the side bar on wide screens.

Comment (harshilsharma63) [reply webapp/src/components/sidebar/sidebar.scss:135]: @kamre That's a UX decision the team took - to only allow hiding the sidebar when the window size is small.

DO you think it's something that should be supported?

Comment (kamre) [reply webapp/src/components/sidebar/sidebar.scss:135]: Yes, in my opinion there are situations when the space is not enough even on wide screens.

Look for example at these kanban boards from default templates, they all have horizontal scroll:
BoardsScroll
This is on laptop with 150% scale in Windows settings.

Comment (harshilsharma63) [reply webapp/src/components/sidebar/sidebar.scss:135]: cc @asaadmahmood

Comment (chenilim) [reply webapp/src/components/sidebar/sidebar.scss:135]: FYI @harshilsharma63, sent PR #1219 to you for review to only hide the collapse-sidebar button for workspaces.
